File: 2025_Oct/youngwon/2025-10-27-Permutation-in-String.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) return false;

        vector<int> s1Count(26, 0), windowCount(26, 0);

        for (char c : s1) s1Count[c - 'a']++;

        int left = 0;
        for (int right = 0; right < s2.size(); right++) {
            windowCount[s2[right] - 'a']++;
            
            if (right - left + 1 == s1.size()) {
                if (windowCount == s1Count) return true;
                
                windowCount[s2[left] - 'a']--;
                left++;
            }
        }

        return false;
    }
};

File: 2025_Oct/Soohyeun/2025-10-27-Permutation-in-String.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_dict = Counter(s1)
        window = Counter(s2[:len(s1)])

        if s1_dict == window:
            return True
        left = 0
        for right in range(len(s1), len(s2)):
            window[s2[left]] -= 1
            if window[s2[left]] == 0:
                del window[s2[left]]
            left += 1
            window[s2[right]] += 1
            if s1_dict == window:
                return True

        return False